File: ravenapp/src/main/java/com/siddhantkushwaha/raven/adapter/MessageAdapter.java
package com.siddhantkushwaha.raven.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.siddhantkushwaha.raven.R;
import com.siddhantkushwaha.raven.localEntity.RavenMessage;
import com.siddhantkushwaha.raven.manager.ThreadManager;
import com.siddhantkushwaha.raven.utility.FirebaseStorageUtil;
import com.siddhantkushwaha.raven.utility.GlideUtilV2;
import com.siddhantkushwaha.raven.utility.JodaTimeUtilV2;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.security.GeneralSecurityException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import io.realm.OrderedRealmCollection;
import io.realm.RealmRecyclerViewAdapter;

public class MessageAdapter extends RealmRecyclerViewAdapter {

    private Context context;
    private OrderedRealmCollection<RavenMessage> data;

    private OnClickListener onClickListener;
    private OnLongClickListener onLongClickListener;

    private FirebaseStorageUtil firebaseStorageUtil;

    public MessageAdapter(Context context, @Nullable OrderedRealmCollection data, boolean autoUpdate) {
        super(data, autoUpdate);

        this.context = context;
        this.data = data;

        this.firebaseStorageUtil = new FirebaseStorageUtil();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        View view = null;
        switch (viewType) {
            case 1:
                view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.layout_message_sent, viewGroup, false);
                return new MessageAdapter.SentMessageHolder(view);
            case 2:
                view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.layout_message_received, viewGroup, false);
                return new MessageAdapter.ReceivedMessageHolder(view);
            default:
                return null;
        }
    }

    @Override
    public int getItemViewType(int position) {
        return data.get(position).getMessageType(FirebaseAuth.getInstance().getUid());
    }

    @Nullable
    @Override
    public RavenMessage getItem(int index) {
        return data.get(index);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

        RavenMessage ravenMessage = data.get(position);

        boolean showDate = true;
        int previousMessageType = -1;
        if (position > 0) {

            RavenMessage previousRavenMessage = data.get(position - 1);

            String currentMessageTime = ravenMessage.getTimestamp();
            if (currentMessageTime == null) currentMessageTime = ravenMessage.getLocalTimestamp();

            String previousMessageTime = previousRavenMessage.getTimestamp();
            if (previousMessageTime == null)
                previousMessageTime = previousRavenMessage.getLocalTimestamp();

            if (JodaTimeUtilV2.dateCmp(DateTime.parse(currentMessageTime), DateTime.parse(previousMessageTime)) == 0)
                showDate = false;

            previousMessageType = this.getItemViewType(position - 1);
        }

        switch (holder.getItemViewType()) {
            case 1:
                ((MessageAdapter.SentMessageHolder) holder).bind(ravenMessage, previousMessageType, showDate, position);
                break;
            case 2:
                ((MessageAdapter.ReceivedMessageHolder) holder).bind(ravenMessage, previousMessageType, showDate, position);
        }
    }

    private class ReceivedMessageHolder extends RecyclerView.ViewHolder {
        View view;
        LinearLayout bannerLayout;
        TextView bannerText;
        ImageView image;
        TextView messageText;
        TextView timeText;
        LinearLayout messageBody;

        ReceivedMessageHolder(View itemView) {
            super(itemView);
            view = itemView;
            bannerLayout = itemView.findViewById(R.id.bannerLayout);
            bannerText = itemView.findViewById(R.id.bannerText);
            image = itemView.findViewById(R.id.image);
            messageText = itemView.findViewById(R.id.text);
            timeText = itemView.findViewById(R.id.sent_time);
            messageBody = itemView.findViewById(R.id.messageBody);
        }

        void bind(RavenMessage ravenMessage, int previousMessageType, boolean showDate, final int position) {

            makeBanner(showDate, bannerLayout, bannerText, ravenMessage);

            setMessageContent(image, messageText, ravenMessage);

            setMessageTime(timeText, ravenMessage);

            setMessageProperties(ravenMessage, messageBody);

            setListener(view, position);
        }
    }

    private class SentMessageHolder extends RecyclerView.ViewHolder {
        View view;
        LinearLayout bannerLayout;
        TextView bannerText;
        ImageView image;
        TextView messageText;
        TextView timeText;
        ImageView status;
        LinearLayout messageBody;

        SentMessageHolder(View itemView) {
            super(itemView);
            view = itemView;
            bannerLayout = itemView.findViewById(R.id.bannerLayout);
            bannerText = itemView.findViewById(R.id.bannerText);
            image = itemView.findViewById(R.id.image);
            messageText = itemView.findViewById(R.id.text);
            timeText = itemView.findViewById(R.id.sent_time);
            status = itemView.findViewById(R.id.status);
            messageBody = itemView.findViewById(R.id.messageBody);
        }

        void bind(RavenMessage ravenMessage, int previousMessageType, boolean showDate, final int position) {

            makeBanner(showDate, bannerLayout, bannerText, ravenMessage);

            setMessageContent(image, messageText, ravenMessage);

            setMessageTime(timeText, ravenMessage);

            if (ravenMessage.getSeenAt() != null)
                status.setBackground(context.getDrawable(R.drawable.badge_message_status_seen));
            else if (ravenMessage.getTimestamp() != null)
                status.setBackground(context.getDrawable(R.drawable.badge_message_status_sent));
            else
                status.setBackground(context.getDrawable(R.drawable.badge_message_status_pending));

            setMessageProperties(ravenMessage, messageBody);

            setListener(view, position);
        }
    }


    public interface OnClickListener {
        void onClickListener(View view, int position);
    }

    public interface OnLongClickListener {
        void onLongClickListener(View view, int position);
    }

    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public void setOnLongClickListener(OnLongClickListener onLongClickListener) {
        this.onLongClickListener = onLongClickListener;
    }

    // Utility functions begin from here .........

    private void makeBanner(boolean showDate, LinearLayout bannerLayout, TextView bannerText, RavenMessage ravenMessage) {

        if (showDate) {

            bannerLayout.setVisibility(View.VISIBLE);
            bannerText.setVisibility(View.VISIBLE);

            String messageTime = ravenMessage.getTimestamp();
            if (messageTime == null)
                messageTime = ravenMessage.getLocalTimestamp();

            bannerText.setText(DateTimeFormat.forPattern("MMMM dd, yyyy").print(DateTime.parse(messageTime)));

        } else {
            bannerLayout.setVisibility(View.GONE);
            bannerText.setVisibility(View.GONE);
        }
    }

    private void setMessageContent(ImageView imageView, TextView messageText, RavenMessage ravenMessage) {

        String text = ravenMessage.getText();
        String fileRef = ravenMessage.getFileRef();

        if (text == null && fileRef == null) {

            messageText.setVisibility(View.VISIBLE);
            setErrorView(messageText, "Message Deleted.");

            imageView.setVisibility(View.GONE);

        } else if (text == null) {

            messageText.setVisibility(View.GONE);
            setDefaultView(messageText, ravenMessage, null);

            imageView.setVisibility(View.VISIBLE);

            // download media
            loadImage(imageView, fileRef);

        } else if (fileRef == null) {

            messageText.setVisibility(View.VISIBLE);
            setMessageText(messageText, ravenMessage);

            imageView.setVisibility(View.GONE);
        } else {

            messageText.setVisibility(View.VISIBLE);
            setMessageText(messageText, ravenMessage);

            imageView.setVisibility(View.VISIBLE);

            // download media
            loadImage(imageView, fileRef);
        }
    }

    private void setMessageText(TextView messageText, RavenMessage ravenMessage) {
        try {
            setDefaultView(messageText, ravenMessage, ThreadManager.decryptMessage(ravenMessage.getThreadId(), ravenMessage.getText()));
        } catch (GeneralSecurityException e) {
            setErrorView(messageText, "Couldn't Decrypt");
        } catch (Exception e) {
            setErrorView(messageText, "There was a problem.");
        }
    }

    private void setMessageTime(TextView timeText, RavenMessage ravenMessage) {
        if (ravenMessage.getTimestamp() != null) {

            timeText.setVisibility(View.VISIBLE);

            DateTime time = DateTime.parse(ravenMessage.getTimestamp());
            DateTimeFormatter build = DateTimeFormat.forPattern("hh:mm a");
            timeText.setText(build.print(time));
        } else {
            timeText.setVisibility(View.GONE);
        }
    }

    private void setMessageProperties(RavenMessage ravenMessage, LinearLayout messageBody) {

        if (ravenMessage.getSelected())
            messageBody.setBackgroundResource(R.color.colorMessageSelected);
        else
            messageBody.setBackgroundResource(android.R.color.transparent);
    }

    private void setListener(View view, Integer position) {
        view.setOnClickListener(v -> {
            if (onClickListener != null)
                onClickListener.onClickListener(v, position);
        });

        view.setOnLongClickListener(v -> {
            if (onLongClickListener != null)
                onLongClickListener.onLongClickListener(v, position);
            return false;
        });
    }

    private void setDefaultView(TextView messageText, RavenMessage ravenMessage, String decryptedMessaged) {

        messageText.setText(decryptedMessaged);
        switch (ravenMessage.getMessageType(FirebaseAuth.getInstance().getUid())) {
            case 1:
                messageText.setTextColor(ContextCompat.getColor(context, R.color.colorWhite));
                messageText.setBackground(context.getDrawable(R.drawable.background_message_holder_indigo));
                break;
            case 2:
                messageText.setTextColor(ContextCompat.getColor(context, R.color.colorBlack));
                messageText.setBackground(context.getDrawable(R.drawable.background_message_holder_white));
                break;
        }
    }

    private void setErrorView(TextView messageText, String errorMessage) {
        messageText.setText(errorMessage);
        messageText.setTextColor(ContextCompat.getColor(context, R.color.colorWhite));
        messageText.setBackground(context.getDrawable(R.drawable.background_message_holder_red));
    }

    private void loadImage(ImageView imageView, String fileRef) {
        firebaseStorageUtil.getDownloadUrl(context, fileRef, uri -> {

            GlideUtilV2.loadImageInChat(context, uri, imageView);
        });
    }
}

File: ravenapp/src/main/java/com/siddhantkushwaha/raven/adapter/MessageAdapter.kt
package com.siddhantkushwaha.raven.adapter

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.siddhantkushwaha.nuttertools.JodaTimeUtil
import com.siddhantkushwaha.raven.R
import com.siddhantkushwaha.raven.manager.ThreadManager
import com.siddhantkushwaha.raven.realm.entity.RavenMessage
import com.siddhantkushwaha.raven.realm.entity.RavenThread
import com.siddhantkushwaha.raven.utility.Common
import com.siddhantkushwaha.raven.utility.FirebaseStorageUtil
import com.siddhantkushwaha.raven.utility.GlideUtilV2
import io.realm.OrderedRealmCollection
import io.realm.RealmRecyclerViewAdapter
import io.realm.RealmResults
import kotlinx.android.synthetic.main.layout_message_banner.view.*
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat
import java.security.GeneralSecurityException
import java.util.*
import kotlin.collections.HashMap

class MessageAdapter(private val context: Context, private val ravenThreadResult: RealmResults<RavenThread>, data: OrderedRealmCollection<RavenMessage>, autoUpdate: Boolean, private val onClickListener: OnClickListener) : RealmRecyclerViewAdapter<RavenMessage, RecyclerView.ViewHolder>(data, autoUpdate) {

    private val colorsForUsers = HashMap<String, Int>()
    private val timeBanner = HashMap<String, String>()

    override fun getItemViewType(position: Int): Int {
        return data!![position].getMessageType(FirebaseAuth.getInstance().uid!!)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            1 -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_message_sent_simple, parent, false)
                return SentMessageViewHolder(this@MessageAdapter, view)
            }

            2 -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_message_sent_image, parent, false)
                return SentMessageWithImageViewHolder(this@MessageAdapter, view)
            }

            3 -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_message_received_simple, parent, false)
                return ReceivedMessageViewHolder(this@MessageAdapter, view)
            }

            4 -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_message_received_image, parent, false)
                return ReceivedMessageWithImageViewHolder(this@MessageAdapter, view)
            }

            -1 -> {
                // TODO TBD
            }
        }

        throw RuntimeException("Message didn't match any category.")
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        val ravenMessage = data?.get(position)!!
        val curTime = ravenMessage.timestamp ?: ravenMessage.localTimestamp

        if (curTime != null && position == 0) {
            timeBanner[ravenMessage.messageId] = curTime
        } else {
            val rvp = data?.get(position - 1)!!
            val preTime = rvp.timestamp ?: rvp.localTimestamp
            if (curTime != null && preTime != null)
                if (JodaTimeUtil.dateCmp(DateTime.parse(curTime), DateTime.parse(preTime)) != 0)
                    timeBanner[ravenMessage.messageId] = curTime
        }

        when (holder.itemViewType) {

            1 -> (holder as SentMessageViewHolder).bind(ravenMessage)
            2 -> (holder as SentMessageWithImageViewHolder).bind(ravenMessage)
            3 -> (holder as ReceivedMessageViewHolder).bind(ravenMessage)
            4 -> (holder as ReceivedMessageWithImageViewHolder).bind(ravenMessage)
        }
    }

    private class SentMessageViewHolder(val messageAdapter: MessageAdapter, itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(ravenMessage: RavenMessage) {
            val text: TextView = itemView.findViewById(R.id.text)
            val time: TextView = itemView.findViewById(R.id.time)

            messageAdapter.setMessageText(text, ravenMessage.text)

            messageAdapter.setMessageTimeAndStatus(time, ravenMessage.timestamp, ravenMessage.isSeenByAll)
            messageAdapter.setProperties(itemView, ravenMessage)
            messageAdapter.setListener(itemView, ravenMessage)
        }
    }

    private class SentMessageWithImageViewHolder(val messageAdapter: MessageAdapter, itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(ravenMessage: RavenMessage) {
            val image: ImageView = itemView.findViewById(R.id.image)
            val text: TextView = itemView.findViewById(R.id.text)
            val time: TextView = itemView.findViewById(R.id.time)

            messageAdapter.setImage(image, ravenMessage.fileRef)

            if (ravenMessage.text != null) {
                text.visibility = View.VISIBLE
                messageAdapter.setMessageText(text, ravenMessage.text)
            } else
                text.visibility = View.GONE

            messageAdapter.setMessageTimeAndStatus(time, ravenMessage.timestamp, ravenMessage.isSeenByAll)
            messageAdapter.setProperties(itemView, ravenMessage)
            messageAdapter.setListener(itemView, ravenMessage)
        }
    }

    private class ReceivedMessageViewHolder(val messageAdapter: MessageAdapter, itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(ravenMessage: RavenMessage) {
            val body: LinearLayout = itemView.findViewById(R.id.body)
            val name: TextView = itemView.findViewById(R.id.name)
            val text: TextView = itemView.findViewById(R.id.text)
            val time: TextView = itemView.findViewById(R.id.time)

            messageAdapter.setName(name, ravenMessage.sentByUserId)

            messageAdapter.setMessageText(text, ravenMessage.text, body)

            messageAdapter.setMessageTimeAndStatus(time, ravenMessage.timestamp)
            messageAdapter.setProperties(itemView, ravenMessage)
            messageAdapter.setListener(itemView, ravenMessage)
        }
    }

    private class ReceivedMessageWithImageViewHolder(val messageAdapter: MessageAdapter, itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(ravenMessage: RavenMessage) {
            val image: ImageView = itemView.findViewById(R.id.image)
            val body: LinearLayout = itemView.findViewById(R.id.body)
            val name: TextView = itemView.findViewById(R.id.name)
            val text: TextView = itemView.findViewById(R.id.text)
            val time: TextView = itemView.findViewById(R.id.time)

            messageAdapter.setName(name, ravenMessage.sentByUserId)

            messageAdapter.setImage(image, ravenMessage.fileRef)

            if (ravenMessage.text != null) {
                body.visibility = View.VISIBLE
                text.visibility = View.VISIBLE
                messageAdapter.setMessageText(text, ravenMessage.text, body)
            } else {
                body.visibility = name.visibility
                text.visibility = View.GONE
            }

            messageAdapter.setMessageTimeAndStatus(time, ravenMessage.timestamp)
            messageAdapter.setProperties(itemView, ravenMessage)
            messageAdapter.setListener(itemView, ravenMessage)
        }
    }


    fun setProperties(itemView: View, ravenMessage: RavenMessage) {

        if (ravenMessage.selected)
            itemView.setBackgroundResource(R.color.colorMessageSelected)
        else
            itemView.setBackgroundResource(android.R.color.transparent)

        setTimeBanner(ravenMessage.messageId, itemView)
    }

    private fun setTimeBanner(messageId: String, itemView: View) {

        val timeText = timeBanner[messageId]

        val insertPoint = itemView.findViewById(R.id.root) as ViewGroup
        val banner = insertPoint.findViewById<View>(R.id.banner)

        if (timeText != null) {
            val text = DateTimeFormat.forPattern("MMMM dd, yyyy").print(DateTime.parse(timeText))
            if (banner != null) {
                banner.visibility = View.VISIBLE
                banner.banner.text = text
            } else {
                val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                val bannerLayout = layoutInflater.inflate(R.layout.layout_message_banner, null)
                bannerLayout.banner.text = text
                insertPoint.addView(bannerLayout, 0, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
            }
        } else if (banner != null) {
            banner.visibility = View.GONE
        }
    }

    fun setImage(imageView: ImageView, fileRef: String?) {

        imageView.setImageDrawable(context.getDrawable(R.drawable.hourglass))
        if (fileRef != null)
            FirebaseStorageUtil.getDownloadUrl(context, fileRef) { url ->
                GlideUtilV2.loadImageInChat(context, url, imageView)
            }
    }

    fun setName(nameTextView: TextView, userId: String?) {

        val ravenThread = ravenThreadResult.first(null)
        if (ravenThread != null && ravenThread.isGroup && userId != null) {

            val ravenUser = ravenThread.users?.findLast { ru ->
                ru.userId == userId
            }

            val name = ravenUser?.contactName ?: ravenUser?.displayName
            ?: ravenUser?.phoneNumber
            ?: "Raven User"


            nameTextView.visibility = View.VISIBLE
            nameTextView.text = name
            nameTextView.setTextColor(getColor(userId))
        } else
            nameTextView.visibility = View.GONE
    }

    private fun getColor(userId: String): Int {

        if (!colorsForUsers.containsKey(userId))
            colorsForUsers[userId] = getRandomColor()

        return colorsForUsers[userId] ?: getRandomColor()
    }

    private fun getRandomColor(): Int {

        val rand = Random()
        return Color.rgb(rand.nextInt(255), rand.nextInt(255), rand.nextInt(255))
    }

    fun setMessageText(text: TextView, encryptedText: String?, body: LinearLayout? = null) {

        val colorStateList = Common.getColorStateList(Color.rgb(163, 163, 163))
        val ravenThread = ravenThreadResult.first(null)
        if (ravenThread != null) {
            try {
                setMessageTextUtil(text, body, ThreadManager.decryptMessage(ravenThread.threadId, encryptedText), Color.WHITE, Color.BLACK, if (body == null) null else Common.getColorStateList(Color.WHITE))
            } catch (e: GeneralSecurityException) {
                setMessageTextUtil(text, body, "Couldn't decrypt.", Color.WHITE, Color.WHITE, colorStateList)
            } catch (e: NullPointerException) {
                setMessageTextUtil(text, body, "Message Deleted.", Color.WHITE, Color.WHITE, colorStateList)
            } catch (e: Exception) {
                setMessageTextUtil(text, body, "There was a problem.", Color.WHITE, Color.WHITE, colorStateList)
            }
        } else
            setMessageTextUtil(text, body, " ", Color.WHITE, Color.BLACK, if (body == null) null else Common.getColorStateList(Color.WHITE))
    }

    private fun setMessageTextUtil(text: TextView, body: LinearLayout?, message: String, color1: Int, color2: Int, colorStateList: ColorStateList?) {

        text.text = message
        text.setTextColor(if (body == null) color1 else color2)
        (body ?: text).backgroundTintList = colorStateList
    }

    fun setMessageTimeAndStatus(time: TextView, timestamp: String?, seenByAll: Boolean? = null) {

        if (timestamp != null) {
            time.visibility = View.VISIBLE
            time.text = DateTimeFormat.forPattern("hh:mm a").print(DateTime.parse(timestamp))

            if (seenByAll != null) {
                if (seenByAll) {
                    time.setTextColor(Color.BLACK)
                    time.backgroundTintList = Common.getColorStateList(Color.rgb(113, 255, 146))
                } else {
                    time.setTextColor(Color.WHITE)
                    time.backgroundTintList = null
                }
            }

        } else
            time.visibility = View.GONE
    }

    fun setListener(itemView: View, ravenMessage: RavenMessage) {

        itemView.setOnClickListener {
            onClickListener.onClick(ravenMessage)
        }

        itemView.setOnLongClickListener {
            onClickListener.onLongClick(ravenMessage)
            true
        }
    }

    interface OnClickListener {
        fun onClick(ravenMessage: RavenMessage)
        fun onLongClick(ravenMessage: RavenMessage)
    }
}